Mathlib/RingTheory/WittVector/Isocrystal.lean

```diff
@@ -0,0 +1,258 @@
+-/
+def Isocrystal.frobenius : V ≃ᶠˡ[p, k] V :=
+  @Isocrystal.frob p _ k _ _ _ _ _ _ _
+#align witt_vector.isocrystal.frobenius WittVector.Isocrystal.frobenius
+
+variable (V)
+
+scoped[Isocrystal] notation "Φ(" p ", " k ")" => WittVector.Isocrystal.frobenius p k
+
+/-- A homomorphism between isocrystals respects the Frobenius map. -/
+-- Porting note(https://github.com/leanprover-community/mathlib4/issues/5171):
+-- removed @[nolint has_nonempty_instance]
+structure IsocrystalHom extends V →ₗ[K(p, k)] V₂ where
+  frob_equivariant : ∀ x : V, Φ(p, k) (toLinearMap x) = toLinearMap (Φ(p, k) x)
+#align witt_vector.isocrystal_hom WittVector.IsocrystalHom
+
+/-- An isomorphism between isocrystals respects the Frobenius map. -/
+-- Porting note(https://github.com/leanprover-community/mathlib4/issues/5171):
+-- removed @[nolint has_nonempty_instance]
+structure IsocrystalEquiv extends V ≃ₗ[K(p, k)] V₂ where
+  frob_equivariant : ∀ x : V, Φ(p, k) (toLinearEquiv x) = toLinearEquiv (Φ(p, k) x)
+#align witt_vector.isocrystal_equiv WittVector.IsocrystalEquiv
+
+scoped[Isocrystal] notation:50 M " →ᶠⁱ[" p ", " k "] " M₂ => WittVector.IsocrystalHom p k M M₂
+
+scoped[Isocrystal] notation:50 M " ≃ᶠⁱ[" p ", " k "] " M₂ => WittVector.IsocrystalEquiv p k M M₂
+
+end PerfectRing
+
+open scoped Isocrystal
+
+/-! ### Classification of isocrystals in dimension 1 -/
+
+
+/-- A helper instance for type class inference. -/
+@[local instance]
+def FractionRing.module : Module K(p, k) K(p, k) :=
+  Semiring.toModule
+#align witt_vector.fraction_ring.module WittVector.FractionRing.module
+
+/-- Type synonym for `K(p, k)` to carry the standard 1-dimensional isocrystal structure
+of slope `m : ℤ`.
+-/
+@[nolint unusedArguments]
+-- Porting note(https://github.com/leanprover-community/mathlib4/issues/5171):
+-- removed @[nolint has_nonempty_instance]
+def StandardOneDimIsocrystal (_m : ℤ) : Type _ :=
+  K(p, k)
+#align witt_vector.standard_one_dim_isocrystal WittVector.StandardOneDimIsocrystal
+
+-- Porting note(https://github.com/leanprover-community/mathlib4/issues/5020): added
+section Deriving
+
+instance : AddCommGroup (StandardOneDimIsocrystal p k m) :=
+  inferInstanceAs (AddCommGroup K(p, k))
+instance : Module K(p, k) (StandardOneDimIsocrystal p k m) :=
+  inferInstanceAs (Module K(p, k) K(p, k))
+
+end Deriving
+
+section PerfectRing
+
+variable [IsDomain k] [CharP k p] [PerfectRing k p]
+
+/-- The standard one-dimensional isocrystal of slope `m : ℤ` is an isocrystal. -/
+instance (m : ℤ) : Isocrystal p k (StandardOneDimIsocrystal p k m) where
+  frob :=
+    (FractionRing.frobenius p k).toSemilinearEquiv.trans
+      (LinearEquiv.smulOfNeZero _ _ _ (zpow_ne_zero m (WittVector.FractionRing.p_nonzero p k)))
+
+@[simp]
+theorem StandardOneDimIsocrystal.frobenius_apply (m : ℤ) (x : StandardOneDimIsocrystal p k m) :
+    Φ(p, k) x = (p : K(p, k)) ^ m • φ(p, k) x := by
+  -- Porting note: was just `rfl`
```

Comment:
I think this deserves some investigation. Why doesn't `rfl` work any more and why is it necessary to use `erw` instead of `rw`.

Reply:
I think it ends up with a different SMul instance? I didn't have much success investigating.